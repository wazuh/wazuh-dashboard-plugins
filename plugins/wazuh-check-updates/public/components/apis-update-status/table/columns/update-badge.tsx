import React, { useState, MouseEvent } from 'react';
import { Update } from '../../../../../common/types';
import {
  EuiBadge,
  EuiDescriptionList,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiSpacer,
  EuiFlexGroup,
  EuiFlexItem,
  EuiLink,
  EuiButton,
} from '@elastic/eui';
import { FormattedMessage, I18nProvider } from '@osd/i18n/react';
import { Markdown } from '../../../../../../../src/plugins/opensearch_dashboards_react/public';
import { getWazuhCore } from '../../../../plugin-services';

export interface UpdateProps {
  update: Update;
}

export const UpdateBadge = ({ update }: UpdateProps) => {
  const { title, description, tag, semver, published_date } = update;

  const {
    utils: { formatUIDate },
  } = getWazuhCore();

  const [isModalVisible, setIsModalVisible] = useState(false);

  const handleOnClickBadge = () => {
    setIsModalVisible(true);
  };

  const handleOnCloseModal = () => {
    setIsModalVisible(false);
  };

  const minorVersion = `${semver.major}.${semver.minor}`;
  const releaseNotesUrl = `https://documentation.wazuh.com/current/release-notes/release-${semver.major}-${semver.minor}-${semver.patch}.html`;
  const upgradeGuideUrl = `https://documentation.wazuh.com/${minorVersion}/upgrade-guide/index.html`;

  return (
    <I18nProvider>
      <>
        <EuiBadge
          color='hollow'
          onClickAriaLabel={`update-${tag}`}
          onMouseDown={(e: MouseEvent<HTMLButtonElement>) => {
            e.stopPropagation();
          }}
          onClick={(e: MouseEvent<HTMLButtonElement>) => {
            e.stopPropagation();
            handleOnClickBadge();
          }}
          style={{ maxWidth: 'max-content' }}
        >
          {tag}
        </EuiBadge>
        {isModalVisible ? (
          <EuiOverlayMask>
            <EuiModal onClose={handleOnCloseModal}>
              <EuiModalHeader>
                <EuiModalHeaderTitle>{title}</EuiModalHeaderTitle>
              </EuiModalHeader>

              <EuiModalBody>
                <EuiDescriptionList
                  listItems={[
                    {
                      title: (
                        <FormattedMessage
                          id={`wazuhCheckUpdates.updateModal.publishedDate`}
                          defaultMessage='Published'
                        />
                      ),
                      description: formatUIDate(new Date(published_date)),
                    },
                  ]}
                />
                <EuiSpacer />
                <EuiFlexGroup responsive={false} wrap>
                  <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
                    <EuiLink href={releaseNotesUrl} target='_blank' external>
                      <FormattedMessage
                        id='wazuhCheckUpdates.updateModal.releaseNotesLink'
                        defaultMessage='Release notes'
                      />
                    </EuiLink>
                  </EuiFlexItem>
                  <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
                    <EuiLink href={upgradeGuideUrl} target='_blank' external>
                      <FormattedMessage
                        id='wazuhCheckUpdates.updateModal.upgradeGuideLink'
                        defaultMessage='Upgrade guide'
                      />
                    </EuiLink>
                  </EuiFlexItem>
                </EuiFlexGroup>
                <EuiSpacer />
                <Markdown markdown={description} openLinksInNewTab />
              </EuiModalBody>

              <EuiModalFooter>
                <EuiButton fill onClick={handleOnCloseModal}>
                  <FormattedMessage
                    id={`wazuhCheckUpdates.updateModal.closeModal`}
                    defaultMessage='Close'
                  />
                </EuiButton>
              </EuiModalFooter>
            </EuiModal>
          </EuiOverlayMask>
        ) : null}
      </>
    </I18nProvider>
  );
};
