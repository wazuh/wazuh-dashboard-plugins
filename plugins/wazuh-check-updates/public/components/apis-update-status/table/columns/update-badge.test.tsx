import React from 'react';
import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { UpdateBadge } from './update-badge';
import userEvent from '@testing-library/user-event';

jest.mock(
  '../../../../../../../node_modules/@elastic/eui/lib/services/accessibility/html_id_generator',
  () => ({
    htmlIdGenerator: () => () => 'htmlId',
  })
);

jest.mock('../../../../utils', () => ({
  formatUIDate: jest.fn().mockReturnValue('2022-05-18T10:12:43Z'),
}));

describe('UpdateBadge component', () => {
  test('should return the UpdateBadge component', () => {
    const { container, getByText } = render(
      <UpdateBadge
        update={{
          description:
            '## Manager\r\n\r\n### Fixed\r\n\r\n- Fixed a crash when overwrite rules are triggered...',
          published_date: '2022-05-18T10:12:43Z',
          semver: {
            major: 4,
            minor: 3,
            patch: 8,
          },
          tag: 'v4.3.8',
          title: 'Wazuh v4.3.8',
        }}
      />
    );

    expect(container).toMatchSnapshot();

    expect(getByText('v4.3.8')).toBeInTheDocument();
  });

  test('should open a modal when click the badge', async () => {
    const { container, getByRole, getByText } = render(
      <UpdateBadge
        update={{
          description:
            '## Manager\r\n\r\n### Fixed\r\n\r\n- Fixed a crash when overwrite rules are triggered...',
          published_date: '2022-05-18T10:12:43Z',
          semver: {
            major: 4,
            minor: 3,
            patch: 8,
          },
          tag: 'v4.3.8',
          title: 'Wazuh v4.3.8',
        }}
      />
    );

    expect(container).toMatchSnapshot();

    const badge = getByRole('button');
    expect(badge).toBeInTheDocument();
    await userEvent.click(badge);
    waitFor(async () => {
      const modalHeader = getByText('Wazuh v4.3.8');
      expect(modalHeader).toBeInTheDocument();

      const modalPublishDate = getByText('2022-05-18T10:12:43Z');
      expect(modalPublishDate).toBeInTheDocument();
    });
  });
});
