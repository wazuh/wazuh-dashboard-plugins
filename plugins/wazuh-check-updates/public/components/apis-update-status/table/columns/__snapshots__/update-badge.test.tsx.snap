// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateBadge component should open a modal when click the badge 1`] = `
<div>
  <button
    aria-label="update-v4.3.8"
    class="euiBadge euiBadge-isClickable euiBadge--hollow euiBadge--iconLeft"
    style="max-width: max-content;"
    title="v4.3.8"
  >
    <span
      class="euiBadge__content"
    >
      <span
        class="euiBadge__text"
      >
        v4.3.8
      </span>
    </span>
  </button>
</div>
`;

exports[`UpdateBadge component should return the UpdateBadge component 1`] = `
<div>
  <button
    aria-label="update-v4.3.8"
    class="euiBadge euiBadge-isClickable euiBadge--hollow euiBadge--iconLeft"
    style="max-width: max-content;"
    title="v4.3.8"
  >
    <span
      class="euiBadge__content"
    >
      <span
        class="euiBadge__text"
      >
        v4.3.8
      </span>
    </span>
  </button>
</div>
`;
